Accelerate StringCoding.hasNegatives for JDK 11, 17, StringCoding.countPositives for JDK 21+ on x86

dylanjtuttle opened the pull request:
This PR accelerates intrinsic candidates `StringCoding.hasNegatives` and `StringCoding.countPositives` on x86, the former on JDK 9-18 and the latter on JDK 19+.

This PR is incremental in a few ways.
- The acceleration currently only delivers the desired performance boost for arrays of up to 31 elements. For arrays of 32 elements and longer, the acceleration still outperforms default OpenJ9, but can no longer keep up with HotSpot. For that, I will need to take advantage of larger SIMD instructions.
- I have discovered a strange performance anomaly in OpenJ9 that causes it to perform significantly faster than expected for `hasNegatives` for arrays of 0-8 elements on JDK 19+. It performs so well for these short arrays that implementing my 'acceleration' would actually cause a performance regression there. While this anomaly is investigated, I will not be accelerating `hasNegatives` on JDK 19+.

In the interest of taking advantage of the performance boost as it currently stands for the 0.51 release, this PR will deliver these changes in their incremental state, with plans for another PR or two down the road to close the aforementioned gaps.

It is also worth noting that this PR is currently not ready to be merged, since I am still investigating a crash on JDK 11 and 17. I am opening this PR anyway to give reviewers as much time as possible before the 0.51 split, since it is rather large.

dylanjtuttle commented:
Paging @vijaysun-omr, @0xdaryl for review and @r30shah, @dchopra001 as requested for comparison to acceleration on Z

vijaysun-omr commented:
Looks fine to me from a quick review. I'll probably defer to @hzongaro for the review since he has much more direct awareness of this work than I do, and can easily review the inliner parts as well (that I skimmed over too).

dylanjtuttle commented:
It appears that the crash is due to a recent OMR commit and is therefore unrelated to my changes. I can build successfully with the `openj9` branch of the `openj9-omr` repo, and can now confirm all of the performance testing checks out.

BradleyWood commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12081,7 +12580,24 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
+               }
+         }
+         break;
+#if JAVA_SPEC_VERSION < 19
```
Excluding this call for JDK < 19 is OK, but why do the same for the method declaration?

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10329,6 +10329,505 @@ inlineCompareAndSwapNative(
+   {
+   // Arguments to countPositives
+   // Byte array
+   auto ba = cg->evaluate(node->getChild(0));
```
I will let you process and action all of Brad's comments before reviewing this thoroughly, but one thing I will mention right away are my strong opinions on the use of the `auto` keyword and its impact on future code understanding for anyone but the author.  I may be known to let one or two slip in on occasion, but since you have many (12) in succession I'll ask that you specify the data type for each (they look to be either `TR::Node *`'s or `TR::Register *`'s in your case.

I'll echo Brad's comments about variable naming as well.  It is easier to read the code later if the variables that contain nodes are suffixed with `Node` and registers are suffixed with `Reg`.

dylanjtuttle commented on runtime/compiler/x/codegen/J9CodeGenerator.cpp:
```diff
@@ -141,6 +141,20 @@ J9::X86::CodeGenerator::initialize()
       }
+#if JAVA_SPEC_VERSION < 19
+   static bool disableInlineStringCodingHasNegatives = feGetEnv("TR_DisableInlineStringCodingHasNegatives") != NULL;
+   if (comp->target().cpu.supportsAVX() && !disableInlineStringCodingHasNegatives &&
```
Do you mean I need to add `&& !TR::Compiler->om.isOffHeapAllocationEnabled()` to these conditions? I was basing this section on other nearby examples and was satisfied by it working until it came time for review. In fact, I don't think this implementation has anything to do with arraylets, so I feel I don't need the `!TR::Compiler->om.canGenerateArraylets()`. What do you think?

dylanjtuttle commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10329,6 +10329,505 @@ inlineCompareAndSwapNative(
+ *   The Code Generator
+ */
+static TR::Register* inlineHasNegatives(TR::Node* node, TR::CodeGenerator* cg)
+   {
```
`countPositives` was introduced in JDK 19, so prior to that `hasNegatives` contains all of that logic and is an intrinsic candidate. This particular PR won't be accelerating `hasNegatives` for JDK 19+, but I was planning to do it later on since I'm already doing it for JDK 9-18 (once I figure out what's going on with the anomaly I mentioned in the PR description).

dylanjtuttle commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10329,6 +10329,505 @@ inlineCompareAndSwapNative(
+   generateRegRegInstruction(TR::InstOpCode::PTESTRegReg, node, xmmchunk, xmmmask, cg);
+
+   // If the result is nonzero (i.e. at least one of the sign bits is set), break and return index
+   generateLabelInstruction(TR::InstOpCode::JNE4, node, returnIndexLabel, cg);
```
The key here was pointed out by Henry early on when I was first trying to figure out how to accelerate `countPositives`:
```
    /**
     * Count the number of leading positive bytes in the range.
     *
     * @implSpec the implementation must return len if there are no negative
     *   bytes in the range. If there are negative bytes, the implementation must return
     *   a value that is less than or equal to the index of the first negative byte
     *   in the range.
     */
```
`countPositives` only needs to return a value less than or equal to the index of the first negative byte. There are a few places in `String.java` that call `countPositives` and then do the work of finding the exact index themselves. I suppose the reason it was designed this way is to take advantage of small time saves in situations where you don't care about the exact index (like when being called by `hasNegatives`).

dylanjtuttle commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10329,6 +10329,505 @@ inlineCompareAndSwapNative(
+    *    if bytes_left > 2               |
+    *       jmp threeOrMoreBytesLabel -+ |
+    *                                  | |
+    *    load 1-2 bytes                | |
```
We have tried a few different residue processing strategies. I haven't actually tried sequential processing, but I have spent a lot of time benchmarking this strategy compared with default OpenJ9 (which, for arrays of 15 elements or less, is essentially doing sequential residue processing). This strategy is based on Annabelle's implementation of `System.arraycopy`, except with the zero length case handled first.

Here are the results of benchmarking `hasNegatives` on JDK 17 (the results are pretty similar across both intrinsics and all JDK versions):

| **# Elements** | **OpenJ9** | **OpenJ9 With Change** | **HotSpot** | **% of HotSpot** |
|:--------------:|:----------:|:----------------------:|:-----------:|:----------------:|
|        0       |    1.11G   |          612M          |     971M    |        63%       |
|        1       |    259M    |          378M          |     554M    |        68%       |
|        3       |    229M    |          356M          |     354M    |        101%       |
|        5       |    194M    |          333M          |     353M    |        94%       |
|        9       |    155M    |          526M          |     299M    |       176%       |
|       15       |    115M    |          531M          |     229M    |       232%       |
|       16       |    114M    |          408M          |     243M    |       168%       |
|       31       |    71.8M   |          353M          |     155M    |       227%       |
|       32       |    71.8M   |          315M          |     485M    |        65%       |
|       63       |    41.3M   |          240M          |     431M    |        56%       |
|       64       |    41.3M   |          211M          |     431M    |        49%       |
|       127      |    22.3M   |          135M          |     350M    |        38%       |
|       128      |    22.6M   |          130M          |     303M    |        43%       |
|       255      |    11.6M   |          85.1M         |     206M    |        41%       |
|       256      |    11.6M   |          81.4M         |     201M    |        40%       |

<img width="616" alt="Screenshot 2025-02-14 at 5 05 02 PM" src="https://github.com/user-attachments/assets/85651cd4-9b41-4f80-a293-64b437885478" />

BradleyWood commented on runtime/compiler/x/codegen/J9CodeGenerator.cpp:
```diff
@@ -143,12 +143,12 @@ J9::X86::CodeGenerator::initialize()
 #if JAVA_SPEC_VERSION < 19
    static bool disableInlineStringCodingHasNegatives = feGetEnv("TR_DisableInlineStringCodingHasNegatives") != NULL;
-   if (comp->target().cpu.supportsAVX() && !disableInlineStringCodingHasNegatives)
+   if (comp->target().cpu.supportsFeature(OMR_FEATURE_X86_SSE2) && !disableInlineStringCodingHasNegatives)
```
The VM cannot run without SSE2, so there is no need to check

dylanjtuttle commented:
@0xdaryl @hzongaro I apparently still have a bug in here somewhere because I'm getting a failure in the JCL during the build, but I thought I'd open up for a second round of review in the meantime while I work on squashing it.

BradleyWood commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10814,11 +10814,15 @@ static TR::Register* inlineHasNegatives(TR::Node* node, TR::CodeGenerator* cg)
    generateLabelInstruction(TR::InstOpCode::label, node, residualTestLabel, cg);
 
+   // Zero out the result register
+   generateRegRegInstruction(TR::InstOpCode::XOR8RegReg, node, resultReg, resultReg, cg);
```
These opcodes wont work on 32-bit. Use `TR::InstOpCode::XORRegReg(is64Bit);` instead. If you really need 64-bits, then disable these intrinsics on 32-bit for now. Check the other instructions for similar issues.

BradleyWood commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -12081,7 +12616,24 @@ J9::X86::TreeEvaluator::directCallEvaluator(TR::Node *node, TR::CodeGenerator *c
-
+      case TR::java_lang_StringCoding_countPositives:
+         {
+         if (comp->target().cpu.supportsFeature(OMR_FEATURE_X86_SSE2) && feGetEnv("replaceCountPositives") != NULL)
```
You already called `cg->setSupportsInlineStringCodingCountPositives()` so you should be calling `cg->getSupportsInlineStringCodingCountPositives()`

dylanjtuttle replied:
I consolidated all of the inlining logic into `InlinerTempForJ9`, since I was already using it to conditionally only inline `countPositives` into `hasNegatives`. I've confirmed that the inlining behaviour is still correct using this method.

BradleyWood commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10329,6 +10329,541 @@ inlineCompareAndSwapNative(
+
+   // limit = offset + length
+   generateRegRegInstruction(TR::InstOpCode::MOV8RegReg, node, limitReg, offsetReg, cg);
+   generateRegRegInstruction(TR::InstOpCode::ADD8RegReg, node, limitReg, lengthReg, cg);
```
Add offset to loopLimit at the end

`loopLimit = (length & -16) + off`

dylanjtuttle commented:
I fixed the bug! I squashed all of my review comment changes into one commit, just doing some more performance testing

hzongaro reviewed (commented):
Great work!  I just have a few suggestions/comments

BradleyWood commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10350,6 +10350,544 @@ inlineCompareAndSwapNative(
+   generateRegRegInstruction(TR::InstOpCode::MOV8RegReg, node, indexReg, offsetReg, cg);
+
+   // limit = offset + length
+   generateRegRegInstruction(TR::InstOpCode::MOV8RegReg, node, limitReg, offsetReg, cg);
```
You still have added offset before ANDing:

```
loopLimit = ((off + length) & -16) + off
```

It should be as follows:

```
loopLimit = (length & -16) + off
```

dylanjtuttle replied:
Hi Brad, I might confused, but I don't think I'm adding offset before ANDing?
```cpp
   // loopLimit = (length & -16) + offset
   generateRegRegInstruction(TR::InstOpCode::MOV4RegReg, node, loopLimitReg, lengthReg, cg);
   generateRegImmInstruction(TR::InstOpCode::AND4RegImm4, node, loopLimitReg, -16, cg);
   generateRegRegInstruction(TR::InstOpCode::ADD4RegReg, node, loopLimitReg, offsetReg, cg);
```

BradleyWood replied:
Sorry, I confused limitReg for loopLimitReg

dylanjtuttle commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10350,6 +10350,544 @@ inlineCompareAndSwapNative(
+   generateLabelInstruction(TR::InstOpCode::label, node, returnFalseLabel, cg);
+
+   // result = 0
+   generateRegImmInstruction(TR::InstOpCode::MOV8RegImm4, node, resultReg, 0, cg);
```
I chose to implement your other suggestion in https://github.com/eclipse-openj9/openj9/pull/21121#discussion_r1970364744 instead, since I combined `inlineHasNegatives` and `inlineCountPositives` and had to choose one.

hzongaro replied:
Yes, I was worried those two suggestions might conflict with one another.  I'll just have another look though.

dylanjtuttle commented:
@0xdaryl I believe I am ready for your review whenever you are!

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10495,6 +10495,306 @@ inlineCompareAndSwapNative(
+   // Deal with the residual (last 15 or fewer) bytes
+   generateLabelInstruction(TR::InstOpCode::label, node, residualLabel, cg);
+
+   // bytesLeft = limit - loopLimit
+   generateRegRegInstruction(TR::InstOpCode::MOV8RegReg, node, bytesLeftReg, limitReg, cg);
+   generateRegRegInstruction(TR::InstOpCode::SUB8RegReg, node, bytesLeftReg, loopLimitReg, cg);
```
`loopLImitReg` isn't used after this point.  I wonder whether it could be used in place of `bytesLeftReg`, like this, to reduce the number of registers needed:

```suggestion
   // Calculate bytes remaining:  loopLimit = -(loopLimit - limit)
   generateRegRegInstruction(TR::InstOpCode::SUB4RegReg, node, loopLimitReg, limitReg, cg);
   generateRegInstruction(TR::InstOpCode::NEG4Reg, node, loopLimitReg, cg);
```

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10495,6 +10495,306 @@ inlineCompareAndSwapNative(
+    */
+
+   // if bytesLeft = 0, jump to returnNoNegativesLabel
+   generateRegImmInstruction(TR::InstOpCode::CMP8RegImm4, node, bytesLeftReg, 0, cg);
```
This can be a four byte comparison

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10495,6 +10495,306 @@ inlineCompareAndSwapNative(
+   generateRegImm64Instruction(TR::InstOpCode::MOV8RegImm64, node, maskReg, 0x8080808080808080, cg);
+
+   // if bytesLeft > 8, jump to nineOrMoreBytesLabel
+   generateRegImmInstruction(TR::InstOpCode::CMP8RegImm4, node, bytesLeftReg, 8, cg);
```
This can be a four byte comparison

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10495,6 +10495,306 @@ inlineCompareAndSwapNative(
+   generateRegRegInstruction(TR::InstOpCode::XOR8RegReg, node, chunkReg, chunkReg, cg);
+
+   // if bytesLeft > 2, jump to threeOrMoreBytesLabel
+   generateRegImmInstruction(TR::InstOpCode::CMP8RegImm4, node, bytesLeftReg, 2, cg);
```
This can be a four byte comparison

hzongaro reviewed (commented):
Just a few more suggestions.

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10474,23 +10475,23 @@ static TR::Register* inlineHasNegativesOrCountPositives(TR::Node* node, TR::Reco
-   // bytesLeft = limit - loopLimit
-   generateRegRegInstruction(TR::InstOpCode::MOV8RegReg, node, bytesLeftReg, limitReg, cg);
-   generateRegRegInstruction(TR::InstOpCode::SUB8RegReg, node, bytesLeftReg, loopLimitReg, cg);
+   // Calculate bytes remaining: lootLimit = -(loopLimit - limit)
```
Typo:  `lootLimit` should be `loopLimit`

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10338,6 +10338,307 @@ inlineCompareAndSwapNative(
+
+   // if index >= loopLimit, jump to handling the residual bytes
+   generateRegRegInstruction(TR::InstOpCode::CMP4RegReg, node, indexReg, loopLimitReg, cg);
+   generateLabelInstruction(TR::InstOpCode::JGE1, node, residualLabel, cg);
```
You should use the 4-byte displacement instructions (i.e., `JGE4`) to be safe.  The binary encoder will automatically convert this to a 1-byte displacement form if it determines the target is within range.

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10338,6 +10338,307 @@ inlineCompareAndSwapNative(
+   else
+      {
+      // Break and calculate index of first negative byte
+      generateLabelInstruction(TR::InstOpCode::JNE1, node, foundNegativeLabel, cg);
```
Use `JNE4` for same reasons as above.

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10338,6 +10338,307 @@ inlineCompareAndSwapNative(
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, indexReg, 16, cg);
+
+   // Jump back to loopLabel
+   generateLabelInstruction(TR::InstOpCode::JMP1, node, loopLabel, cg);
```
`JMP4`

0xdaryl replied:
This applies to all branches in this PR.  I won't comment on each.

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10338,6 +10338,307 @@ inlineCompareAndSwapNative(
+    */
+
+   // if loopLimit = 0, jump to returnNoNegativesLabel
+   generateRegImmInstruction(TR::InstOpCode::CMP4RegImm4, node, loopLimitReg, 0, cg);
```
Should be `TEST4RegReg loopLimitReg, loopLimitReg` for instruction economy

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10338,6 +10338,307 @@ inlineCompareAndSwapNative(
+    *       jmp residualTestLabel        |
+    *                                    |
+    *    nineOrMoreBytesLabel: <---------+
+    *       load 9-16 bytes
```
You should round out the pseudo-code with the rest of the residue handing, otherwise it is incomplete.

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10338,6 +10338,307 @@ inlineCompareAndSwapNative(
+   if (recognizedMethod == TR::java_lang_StringCoding_hasNegatives)
+      {
+      // Return false
+      generateRegRegInstruction(TR::InstOpCode::XOR8RegReg, node, indexReg, indexReg, cg);
```
`XOR4RegReg` for instruction economy.

0xdaryl commented:
Can you attach two logs with traceCG: one that shows your `countPositives` sequence and another for `countNegatives`?  I'd like to review the sequences when it is all laid out.

0xdaryl commented:
Is there a unit test that exercises the various lengths being specialized for `countPositives` and `hasNegatives`?  In particular, I want to confirm that you checked each residue length (0-15) for correctness.

0xdaryl commented:
This intrinsification assumes the arrays are contiguous.  You should prevent this optimization if discontiguous arrays are possible by the GC mode.

dylanjtuttle commented:
@0xdaryl Does that mean to disable the optimization if `comp->om.canGenerateArraylets()`?

0xdaryl commented:
> Does that mean to disable the optimization if comp->om.canGenerateArraylets()?

Yes

dylanjtuttle commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -6002,6 +6002,33 @@ TR_InlinerFailureReason
       return Recognized_Callee;
       }
 
```
Something that just occurred to me about this section: Since this file is common, do I need to guard this logic for x86 only?

0xdaryl replied:
yes, if these decisions only apply to that platform (i.e., do all backends support this intrinsification?  do these 32/64-bit and JDK19+ decisions apply to all platforms?  etc.)

dylanjtuttle commented:
I included all of the tracelogs generated to cover all of the compilations:
[log_CG_countPositives.txt](https://github.com/user-attachments/files/19095167/log_CG_countPositives.txt)
[log_CG_countPositives_2.txt](https://github.com/user-attachments/files/19095169/log_CG_countPositives_2.txt)
[log_CG_hasNegatives.txt](https://github.com/user-attachments/files/19095165/log_CG_hasNegatives.txt)
[log_CG_hasNegatives_2.txt](https://github.com/user-attachments/files/19095166/log_CG_hasNegatives_2.txt)

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10338,6 +10338,317 @@ inlineCompareAndSwapNative(
+   generateLabelInstruction(TR::InstOpCode::JNE4, node, returnHasNegativesLabel, cg);
+
+
+   // returnNoNegatives label
```
A general comment throughout this PR.  You shouldn't comment trivially obvious code, or use comments that essentially repeat what the code says equally or better (another example: "Jump to residualTestLabel").

The instruction sequence being laid out is already long and branchy, and fewer unnecessary comments to fill up a reader's context buffer improves readability immensely.

hzongaro replied:
I'm going to mark this comment as "Unresolved".  I just wanted @0xdaryl to have a chance to verify whether he's satisfied with the comments.

dylanjtuttle commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10338,6 +10338,317 @@ inlineCompareAndSwapNative(
+   else
+      {
+      // result = index - offset
+      generateRegRegInstruction(TR::InstOpCode::SUB4RegReg, node, indexReg, offsetReg, cg);
```
Yep, you've got it. I'll add a little loop to find the exact index.

BradleyWood replied:
I think no matter what JCL spec says, the compiler should never change the behaviour of a method.

You do not need a loop to do this. You have a non-zero bit mask. Use tzcnt instruction to find the index.

dylanjtuttle replied:
It hadn't occurred to me that I could use tzcnt here as well. I don't know exactly how tzcnt works under the hood but I assume it works faster than a loop would. This would save us time in the callers to `countPositives` who usually loop from whatever index is returned from `countPositives` to the actual index of the first negative byte at the presumably low cost of two extra instructions of overhead inside `countPositives`. That seems like a pretty fair trade to me.

BradleyWood replied:
Just a disclamer, you made a lot of changes since I last look at this PR in depth. But, if you have a bitmask, such as one from movmsk instruction, and you know its non-zero, you can find the index of the first or last '1' bit using leading/trailing zero count instruction.

0xdaryl replied:
I have my doubts that a single instruction could help you here (in the residue part) based on the approach you're currently using.  For example, in the case of `length > 8`, you essentially `OR` an overlapping lower "chunk" and an upper "chunk" and then test the union with your mask (0x8080...).  How will you properly distinguish between a negative value at index 7 and a negative value at index 8 since they will be served by the same part of the mask?

One thing to keep in mind for the residue computation is that as it becomes increasingly complex, you have to trade the performance of that agains the simplicity of just simply iterating through the (max) 15 bytes and checking for negatives (just as the JCL is doing).

dylanjtuttle replied:
Good call, I had just noticed that as I was working on implementing the tzcnt idea. We could potentially rework the residual section to load into another xmm register, mask out the garbage bytes at the end, and then treat it like we do in the loop. Alternatively, we could just iterate through the up to 15 bytes like you mentioned.

hzongaro replied:
> I think no matter what JCL spec says, the compiler should never change the behaviour of a method.

@dylanjtuttle took this approach based on my recommendation.  The `StringCoding.countPositives` method is marked with the `IntrinsicCandidate` annotation, and I based that advice on a superficial understanding of what `IntrinsicCandidate` signified.  In particular, I thought that a JIT compiler could choose to replace the implementation with something equivalent that performed the required functionality &mdash; namely, returning an index at or before the first negative byte, if the byte array contains any negative bytes.

However, on reading JEP 348[1] more carefully I believe that that replacement would have to happen at the level of the `javac` compiler, perhaps with some alternative bytecodes invoking a different, specialized implementation of `countPositives` in place of the usual method invocation of `StringCoding.countPositives`.  If that specialized implementation returned an index at or before the first negative byte, the bytecode interpreter and JIT-compiled code would produce identical results, and that would be acceptable.

As OpenJ9 is still invoking the original implementation of `StringCoding.countPositives`, the result produced by JIT-compiled code should match the result that would be produced by the bytecode interpreter, but it won't as this pull request stands.  So that likely means that you'll need to have the residual byte handling find the index of the actual first negative byte, as @0xdaryl pointed out, and leverage that residual byte handling code to deal with the case where a negative byte was found inside one of the sixteen byte chunks that the main loop processes.

@dylanjtuttle, my apologies for leading you astray!

[1] https://openjdk.org/jeps/348

dylanjtuttle replied:
Understood, I'll work on that!

hzongaro replied:
@BradleyWood pointed out off-line that JEP 348 was withdrawn.  The specification of the `IntrinsicCandidate` annotation is a little unclear, but it seems safest to assume that the implementation the JIT uses should match the behaviour that the JVM’s bytecode interpreter uses, which is the actual bytecode implementation of `StringCoding.countPositives`.

0xdaryl commented:
Apologies for piling on the comments asynchronously, but another thing occurs to me.

In the JCL implementation, it is possible for the `ba` array to be NULL or for accesses to `ba` to produce an AIOB.  You dont handle those cases in your intrinsic.  To guard against that, you'd have to wrap the intrinsification with a check for a NULL `ba` and that the size of `ba` is within the length specified.  If either of those cases could happen, then just call the helper.  This is similar to what Konno-san did for the `StringLatin1.inflate` intrinsic.

BradleyWood commented:
> In the JCL implementation, it is possible for the ba array to be NULL 

Its true, but I think its not needed for known caller classes like String. I think in most (probably all) calls to countPositives in String class are protected by a call to `checkBoundsOffCount(offset, length, bytes.length)` and `bytes.length` should already have a null check.

dylanjtuttle commented:
I have implemented a simple iterative routine for handling the residual bits. Now that I have a unit test that checks all of the cases, I can confirm that it is correct. I will be doing some performance testing to see how this change affects our standing.

dylanjtuttle commented:
After doing some performance testing, the iterative approach is not going to cut it. I will try my xmm register + masking garbage bytes approach instead.

dylanjtuttle commented:
I have implemented a residue handling algorithm based on the original implementation, but with added logic to find the exact index. The code is getting pretty big and unwieldy, and I still need to do performance testing tomorrow, but it does work and it should be noticeably faster than the iterative method, which is a good place to build from at least.

dylanjtuttle commented:
Adding the exact index logic slows us down a little, but the performance is certainly still satisfactory, far outperforming OpenJ9. @hzongaro could I get a review on this new version?

hzongaro commented:
@dylanjtuttle, sorry for the late request, but may I ask you to attach updated versions of the log files that you posted in an [earlier comment](#issuecomment-2701841441)?

dylanjtuttle commented:
@hzongaro Done! Here they are:

[log_countPositives_exactIndex_1.txt](https://github.com/user-attachments/files/19291339/log_countPositives_exactIndex_1.txt)
[log_countPositives_exactIndex_2.txt](https://github.com/user-attachments/files/19291340/log_countPositives_exactIndex_2.txt)
[log_hasNegatives_exactIndex.txt](https://github.com/user-attachments/files/19291341/log_hasNegatives_exactIndex.txt)

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10341,6 +10341,473 @@ inlineCompareAndSwapNative(
+      generateLabelInstruction(TR::InstOpCode::JE4, node, threeOrMoreSecondLoadLabel, cg);
+
+      // Otherwise, count the number of zeroes in the chunk before the first sign bit
+      generateRegRegInstruction(TR::InstOpCode::TZCNT2RegReg, node, chunkReg, chunkReg, cg);
```
Should this be using `lzcnt` instead?  I think the main loop that uses `pmovmskb` needs to use `tzcnt` operations because of the order in which the bits are placed in the target register.  In this case, I think you want to know the position of the first sign bit, whereas `tznct` will give you the position of the last sign bit.

dylanjtuttle replied:
Since x86 is little endian (as far as I can tell that's something that we can always rely on), `tzcnt` does actually find the first sign bit, I've tested it out to make sure.

hzongaro replied:
Oops!  I hadn't been thinking about the fact that those bytes would be reversed in loading them into a register.

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10341,6 +10341,473 @@ inlineCompareAndSwapNative(
+      generateLabelInstruction(TR::InstOpCode::JE4, node, returnNoNegativesLabel, cg);
+
+      // Otherwise, count the number of zeroes in the chunk before the first sign bit
+      generateRegRegInstruction(TR::InstOpCode::TZCNT4RegReg, node, chunkReg, chunkReg, cg);
```
Again, should this be `lzcnt` rather than `tzcnt`?

hzongaro replied:
My bad.

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10341,6 +10341,473 @@ inlineCompareAndSwapNative(
+      generateLabelInstruction(TR::InstOpCode::JE4, node, returnNoNegativesLabel, cg);
+
+      // Otherwise, count the number of zeroes in the chunk before the first sign bit
+      generateRegRegInstruction(TR::InstOpCode::TZCNT2RegReg, node, chunkReg, chunkReg, cg);
```
Again, should this be `lzcnt`?

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10341,6 +10341,473 @@ inlineCompareAndSwapNative(
+      generateLabelInstruction(TR::InstOpCode::JE4, node, fiveOrMoreSecondLoadLabel, cg);
+
+      // Otherwise, count the number of zeroes in the chunk before the first sign bit
+      generateRegRegInstruction(TR::InstOpCode::TZCNT4RegReg, node, chunkReg, chunkReg, cg);
```
One more, `lzcnt`?

hzongaro replied:
No, because the machine is little-endian, as @dylanjtuttle pointed out.

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10341,6 +10341,473 @@ inlineCompareAndSwapNative(
+   generateRegRegInstruction(TR::InstOpCode::SUB4RegReg, node, loopLimitReg, limitReg, cg);
+   generateRegInstruction(TR::InstOpCode::NEG4Reg, node, loopLimitReg, cg);
+
+   /*
```
Thinking about the handling of the residue bytes for `countPositives`, I wonder whether it could be simplified by doing something like this.

If there are eight or more bytes remaining, load eight bytes, mask the sign bits, and use `lzcnt` to check for negative bytes.  If any were found, update the index, and finish.

For the remaining seven or fewer bytes, load them in chunks, conditionally, based on the number of bytes remaining - four, then two, then one - then use `lzcnt` on all of them.  I can sketch this out in more detail if it's not clear.

dylanjtuttle replied:
I think I see what you're saying! By "using `lzcnt` on all of them", do you mean loading all of those bytes into the same register and then using `lzcnt` afterwards? Would that involve shifting after each load to line everything up correctly?

dylanjtuttle replied:
I have implemented this processing strategy by first moving to the end of the residual bytes and then working backwards, first possibly loading the last 4 bytes, then 2 (shifting if necessary), then 1 (again, shifting if necessary). This places all of the residual bytes in the correct order in the chunk register, allowing us to calculate the index of any negative bytes.

hzongaro reviewed (commented):
Thanks, Dylan.  I have some suggestions about the handling of the residue bytes for `countPositives`.

dylanjtuttle commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10341,6 +10341,473 @@ inlineCompareAndSwapNative(
+   // Load 16 bytes from address [buf + index]
+   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, xmmChunkReg, generateX86MemoryReference(bufReg, indexReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg), cg);
+
+   // Create zero vector
+   generateRegRegInstruction(TR::InstOpCode::PXORRegReg, node, xmmMaskReg, xmmMaskReg, cg);
+
+   // Compare 0 > chunk bytewise
+   generateRegRegInstruction(TR::InstOpCode::PCMPGTBRegReg, node, xmmMaskReg, xmmChunkReg, cg);
+
+   // Extract bitmask of negative values
+   generateRegRegInstruction(TR::InstOpCode::PMOVMSKB4RegReg, node, maskReg, xmmMaskReg, cg);
```
You're right! I've removed those redundant instructions.

hzongaro replied:
Just looking at this again, are the `PXORRegReg` and `PCMPGTBRegReg` instructions in this sequence still needed?  Couldn't the `PMOVMSKB4RegReg` instruction be applied directly to `xmmChunkReg`?

dylanjtuttle replied:
Good catch! A casualty of reverting back to the inexact implementation that I missed.

dylanjtuttle commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10387,6 +10387,328 @@ inlineCompareAndSwapNative(
+   // Otherwise, load 8 bytes at address [buf + index] into the chunk register
+   generateRegMemInstruction(TR::InstOpCode::L8RegMem, node, chunkReg, generateX86MemoryReference(bufReg, indexReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg), cg);
+
+   // Increment the index by 8, in preparation for the case that this chunk has no negative bytes and we move on to the sevenOrFewerBytes case
+   generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, indexReg, 8, cg);
+
+   // AND the chunk register with the mask
+   generateRegRegInstruction(TR::InstOpCode::AND8RegReg, node, chunkReg, maskReg, cg);
+   // If the result is zero (meaning none of the bytes are negative), jump to sevenOrFewerBytesLabel
+   generateLabelInstruction(TR::InstOpCode::JE4, node, sevenOrFewerBytesLabel, cg);
+
+   // Otherwise, we have found at least one negative byte, so decrement the index again
+   generateRegImmInstruction(TR::InstOpCode::SUB4RegImm4, node, indexReg, 8, cg);
```
I wasn't a huge fan of the way I handled this section, but I couldn't think of a way to do it without some amount of waste.

In this section, we're handling the case in which there are 8 or more residual bytes. To handle this case, we load the first 8 bytes and check for any negatives. If there are no negatives, we increment the index by 8 and move on to the `sevenOrFewerBytes` case. If we do find any negatives, we don't increment the index because we need to use it to calculate the index of that first negative byte.

The problem is that the `sevenOrFewerBytes` case doesn't know whether there were ever more residual bytes, or more specifically, whether we did that 8 byte load before it that required incrementing the index. All it knows is that there are currently 7 or fewer unhandled residual bytes remaining, starting at whatever byte the index is currently pointing to. We need to increment the index at a place in the code that knows whether that 8 byte load took place or not, so that means either incrementing unconditionally and then decrementing again if we fall through to the case where there were negative bytes after all, or setting some kind of boolean that can tell us whether to increment once we get to the `sevenOrFewerBytes` case. Operating under the assumption that we should avoid conditional branches when reasonably possible, I chose the former, figuring that adding two unnecessary instructions to the (less common) 'does has negatives' case would be better than adding more conditional branching overhead to the (more common) 'does not have negatives' case.

Please anyone let me know if I missed a less wasteful way to do this!

dylanjtuttle commented:
Updated logs for the shift-based residual processing algorithm:

[log_countPositives_shift.txt](https://github.com/user-attachments/files/19455255/log_countPositives_shift.txt)
[log_hasNegatives_shift.txt](https://github.com/user-attachments/files/19455256/log_hasNegatives_shift.txt)

dylanjtuttle commented:
Reverted to non-exact residual handling. Logs:

[log_countPositives_nonExact.txt](https://github.com/user-attachments/files/19592520/log_countPositives_nonExact.txt)
[log_hasNegatives_nonExact.txt](https://github.com/user-attachments/files/19592521/log_hasNegatives_nonExact.txt)

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10387,6 +10387,301 @@ inlineCompareAndSwapNative(
+   // Examine the chunk register now that all of the residual bytes have been ORed into it
+   generateLabelInstruction(TR::InstOpCode::label, node, residualTestLabel, cg);
+   // AND the residual bytes with the new mask
+   generateRegRegInstruction(TR::InstOpCode::TEST8RegReg, node, chunkReg, maskReg, cg);
+   // If the result is nonzero (i.e. at least one of the sign bits is set), jump to returnHasNegativesLabel
+   generateLabelInstruction(TR::InstOpCode::JNE4, node, returnHasNegativesLabel, cg);
+
+
+   // Set result in the case that no negative bytes were found
+   generateLabelInstruction(TR::InstOpCode::label, node, returnNoNegativesLabel, cg);
+   if (recognizedMethod == TR::java_lang_StringCoding_hasNegatives)
+      {
+      // result = false
+      generateRegRegInstruction(TR::InstOpCode::XOR4RegReg, node, indexReg, indexReg, cg);
+      }
```
Sorry - I think I inverted the logic.  I think that should be `setne/SETNE1Reg` instead.

Also, in case it wasn't clear, labels L0072 and L0073 can be combined into a single label that precedes the `setne`.

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10387,6 +10387,301 @@ inlineCompareAndSwapNative(
+   // If we found any negative bytes, calculate the index of the first negative byte in the chunk
+   if (recognizedMethod == TR::java_lang_StringCoding_countPositives)
+      {
+      generateLabelInstruction(TR::InstOpCode::label, node, foundNegativeLabel, cg);
+
+      // Calculate the index of the first negative byte within the chunk
+      generateRegRegInstruction(TR::InstOpCode::TZCNT4RegReg, node, maskReg, maskReg, cg);
+
+      // Add the index, which points to the beginning of the chunk,
+      // for the overall index of the first negative byte
+      generateRegRegInstruction(TR::InstOpCode::ADD4RegReg, node, indexReg, maskReg, cg);
+
+      generateLabelInstruction(TR::InstOpCode::JMP4, node, returnHasNegativesLabel, cg);
```
Given that you're going with the inexact implementation, I think these instructions are no longer needed.  Then the `jne4` to `foundNegativeLabel` above can branch directly to `returnHashNegativesLabel`  instead.

dylanjtuttle commented:
I seem to have a couple of problems at the moment. For some reason, in cases with negative bytes, my implementation seems to be unable to return an index larger than 128. I checked, and this behaviour appeared in my previous commit as well, so it wasn't introduced by my latest changes. There don't appear to be any relevant instructions constraining the size of the index register, so I'm not sure what could be causing this. It still satisfies the requirement of returning a value less than or equal to the index of the first negative, so it's not strictly incorrect, but obviously we should be able to do better than that.

I'm also getting a rather baffling JCL crash relating to an invalid CEN header during the build, which usually indicates I have a bug somewhere, but my unit tests pass correctly (aside from the 128 byte bug mentioned above that still satisfies the spec) and nothing seems out of the ordinary when running the debugger. I've got a Jenkins build running right now to try and eliminate any external factors that may be affecting things in my dev environment.

dylanjtuttle commented:
Thanks to some help from @hzongaro, we figured out that the max index of 128 was also due to a bug in the unit test, which means we are all clear! In that case, @0xdaryl could you give this another review when you have a moment?

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10387,6 +10387,278 @@ inlineCompareAndSwapNative(
+   generateRegMemInstruction(TR::InstOpCode::MOVDQURegMem, node, xmmChunkReg, generateX86MemoryReference(bufReg, indexReg, 0, TR::Compiler->om.contiguousArrayHeaderSizeInBytes(), cg), cg);
+
+   // Extract bitmask of sign bits
+   generateRegRegInstruction(TR::InstOpCode::PMOVMSKB4RegReg, node, maskReg, xmmMaskReg, cg);
```
I think this should be using `xmmChunkReg` instead.  Looking at the code that results during instruction selection, we have this:

```
[0x78f725164560]	movdqu	VRF_0088, xmmword ptr [&GPR_0080+1*GPR_0087+0x8]		# MOVDQURegReg, SymRef [#440 +8]
 [0x78f7251645f0]	pmovmskb	GPR_0086, VRF_0085		# PMOVMSKB4RegReg
 ```
 
 but after register assignment
 
 ```
  [0x78f725164560]	movdqu	xmm0, xmmword ptr [r8+1*rax+0x8]		# MOVDQURegReg, SymRef [#440 +8]
 [0x78f7251645f0]	pmovmskb	?si?, xmm0		# PMOVMSKB4RegReg
 ```
 
 So it looks like you got lucky, and it happened to assign the same vector register for both instructions.
 
 Do you know what is the significance of the question marks around the `?si?`?

dylanjtuttle replied:
Good catch, thank you! In fact, there's no need for an `xmmMaskReg` at all anymore.

hzongaro replied:
I found code in [`TR_Debug::getName(uint32_t realRegisterIndex, TR_RegisterSizes size)`](https://github.com/eclipse-omr/omr/blob/707c74b325df68f43fcf6d720ecbdefbdfe20f41/compiler/x/codegen/X86Debug.cpp#L2123) that prints that register as `?si?` if it doesn’t know the size of the register.  I would recommend looking into how that comes to be to make sure it doesn’t represent a problem.

dylanjtuttle replied:
At Brad's suggestion, I added the property `IA32OpProp_IntTarget` to the `PMOVMSKB4RegReg` instruction in [`omr/compiler/x/codegen/X86Ops.ins`](https://github.com/eclipse-omr/omr/blob/master/compiler/x/codegen/X86Ops.ins#L3618-L3624), and a jitlog confirms that we're now generating a register properly.
```
 [0x789e70562360]	movdqu	xmm0, xmmword ptr [r11+1*rax+0x8]		# MOVDQURegReg, SymRef [#440 +8]
 [0x789e705623f0]	pmovmskb	edi, xmm0		# PMOVMSKB4RegReg
```
The size of the output of `pmovmskb` does vary depending on the size of the input, but it doesn't get any larger than 32 bits and according to the manual it does zero extend in shorter cases, so this seems like a pretty reasonable fix. Should I open a PR against OMR where we can discuss the more broad impact of this change?

hzongaro reviewed (commented):
Sorry - I had a couple of late comments for issues I just noticed.  I should be able to approve after you have addressed these, and after you have squashed commits &mdash; presumably after @0xdaryl completes his review.

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10570,6 +10570,274 @@ inlineCompareAndSwapNative(
+   TR::Register *xmmChunkReg = cg->allocateRegister(TR_VRF);
+   TR::Register *chunkReg = cg->allocateRegister();
+
+   TR::RegisterDependencyConditions *dependencies = generateRegisterDependencyConditions((uint8_t)7, (uint8_t)7, cg);
```
You need to call `stopAddingConditions()` to "close" the dependency list.

Also, if you don't plan on adding anymore deps to this after this point you should just create it with 6 deps rather than 7.

0xdaryl commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10570,6 +10570,274 @@ inlineCompareAndSwapNative(
+   else
+      {
+      generateLabelInstruction(TR::InstOpCode::JE4, node, returnNoNegativesLabel, cg);
+      }
```
Replace with:

```
generateLabelInstruction(TR::InstOpCode::JE4, node,
                         (recognizedMethod == TR::java_lang_StringCoding_hasNegatives) ? returnBooleanLabel : returnNoNegativesLabel, 
                         cg);
```

to save some code.

0xdaryl commented on runtime/compiler/optimizer/InlinerTempForJ9.cpp:
```diff
@@ -6099,6 +6099,49 @@ TR_InlinerFailureReason
       return Recognized_Callee;
       }
 
+   // This logic that decides whether to inline or accelerate hasNegatives and countPositives only applies to x86
```
Thanks for adding that comment, but I think you further need to explain why it only applies to x86.  I hadn't looked at this code for a few weeks and it was already lost on me why that's the case.

dylanjtuttle commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10341,6 +10341,473 @@ inlineCompareAndSwapNative(
+   generateLabelInstruction(TR::InstOpCode::JG4, node, nineOrMoreBytesLabel, cg);
+
+   // Zero out the chunk register
+   generateRegRegInstruction(TR::InstOpCode::XOR8RegReg, node, chunkReg, chunkReg, cg);
```
I think this needs to be a 64 bit instruction because down in the `residualTest` section we're performing a `TEST8RegReg` on the chunk and mask registers to account for any of the cases from 1 byte to 15.

0xdaryl replied:
XOR'ing a reg with itself has the effect of zeroing out the register.  On x86-64, XOR4 will implicitly zero out the upper 32-bits, and has the potential of being a one byte shorter instruction.

dylanjtuttle replied:
Oh I see, thanks!

0xdaryl commented:
See line endings CI failures

dylanjtuttle commented:
At @BradleyWood's suggestion, I've added SIMD encoding for `PMOVMSKB` to specify that we want the 16 byte version. If that encoding isn't supported on the architecture, I've reverted to a version that runs the loop with an 8 byte chunk instead of a 16 byte chunk.

hzongaro commented:
@dylanjtuttle, may I ask you to squash your commits to a reasonable set, and then I will run PR testing.

dylanjtuttle commented:
@hzongaro Done, thanks!

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10570,6 +10570,310 @@ inlineCompareAndSwapNative(
+
+   // If the 16 byte encoding of the pmovmskb instruction is not supported on this architecture,
+   // Prepare an 8 byte sign bit mask so we can run an alternate version of the algorithm
+   if (pmovmskbEncoding == OMR::X86::Encoding::Bad)
```
May I ask you to introduce a `bool` variable to give this condition a name &mdash; `useEightByteChunks` or `useVectorRegister`, say &mdash; and then use that variable in this test and the remaining tests below?

It might also be useful to introduce an environment variable that can be used to force the version that does not use `pmovmskb` to be generated.

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10570,6 +10570,310 @@ inlineCompareAndSwapNative(
+   TR::Register *limitReg = cg->allocateRegister();
+   TR::Register *maskReg = cg->allocateRegister();
+   TR::Register *indexReg = cg->allocateRegister();
+   TR::Register *xmmChunkReg = cg->allocateRegister(TR_VRF);
```
Allocation of the `xmmChunkReg` should be guarded by the test of `pmovmskbEncoding`.  No need to allocate that register and add pre- and post-conditions for it in the case where it's not needed.

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10570,6 +10570,310 @@ inlineCompareAndSwapNative(
+      // Check if any negative values exist
+      generateRegRegInstruction(TR::InstOpCode::TEST2RegReg, node, maskReg, maskReg, cg);
+
+      // If the result is nonzero, we found at least one negative byte
+      generateLabelInstruction(TR::InstOpCode::JNE4,
+                               node,
+                               (recognizedMethod == TR::java_lang_StringCoding_hasNegatives) ? returnBooleanLabel : returnHasNegativesLabel,
+                               cg);
+
+      // increment index by 16 and jump back to the top of the loop
+      generateRegImmInstruction(TR::InstOpCode::ADD4RegImm4, node, indexReg, 16, cg);
+      generateLabelInstruction(TR::InstOpCode::JMP4, node, loopLabel, cg);
```
These lines are the same in both the if and else branches.  I think they can be factored out of the if entirely.

dylanjtuttle replied:
Good point! We increment the index by 16 in the vectorized case and by 8 in the non-vectorized case, but I can still pull them out and add a ternary operator to pick the amount we increment by.

hzongaro reviewed (commented):
Sorry for the late comments.  I hadn't reviewed your most recent changes for testing whether `pmovmskb` instruction is available before.

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10590,26 +10590,43 @@ static TR::Register* inlineHasNegativesOrCountPositives(TR::Node* node, TR::Reco
+   // We need to specify the encoding of this instruction so that the code generator knows we want the 16 byte version
+   TR::InstOpCode pmovmskb = TR::InstOpCode::PMOVMSKB4RegReg;
+   OMR::X86::Encoding pmovmskbEncoding = pmovmskb.getSIMDEncoding(&cg->comp()->target().cpu, TR::VectorLength128);
+   bool useVectorInstructions = (pmovmskbEncoding != OMR::X86::Encoding::Bad) && feGetEnv("TR_disableSIMDHasNegativesCountPositives") == NULL;
```
Code in the JIT typically will assign the result of a call to `feGetEnv` to a static variable so that it will be evaluated only once during any JVM instance.

hzongaro replied:
```suggestion
   static bool disableSIMDHasNegativesCountPositives = feGetEnv("TR_disableSIMDHasNegativesCountPositives") != NULL
   bool useVectorInstructions = (pmovmskbEncoding != OMR::X86::Encoding::Bad) && !disableSIMDHasNegativesCountPositives;
```

hzongaro reviewed (approved):
Looks good.  If you go ahead and squash these latest commits again, I will start the testing.  Thanks!

hzongaro commented:
Jenkins test sanity.functional,sanity.openjdk win,win32,xlinux,xmac jdk8,jdk11,jdk17,jdk21

dylanjtuttle commented:
It looks like we're getting failures on x86 Mac in some string manipulation tests, a SIGBUS in `ZipFile.java` (which I'm pretty sure I've seen failures in when debugging), and a bunch of messages that seem to allude to 128 bit vectorization. This is most likely related to my changes. @hzongaro I wonder if this could be caused by the missing property in the `pmovmskb` instruction. Would it make sense to try running these tests again now that that change has been merged to see if that fixes things?

hzongaro commented:
Rerunning testing with OMR change

Jenkins test sanity.functional,sanity.openjdk win,win32,xlinux,xmac jdk8,jdk11,jdk17,jdk21 depends eclipse-omr/omr#7746

BradleyWood commented:
If you have a log of your intrinsic during any of your test failures I would be happy to take a look. @dylanjtuttle

hzongaro reviewed (approved):
I think your latest changes look good.  Once you have squashed the commits, I will kick off PR testing.

hzongaro commented on runtime/compiler/x/codegen/J9TreeEvaluator.cpp:
```diff
@@ -10570,6 +10570,340 @@ inlineCompareAndSwapNative(
+   {
+   // Arguments to hasNegatives or countPositives
+   // Byte array
+   TR::Register *bufReg = cg->longClobberEvaluate(node->getChild(0));
```
If I understand correctly, when a "clobber evaluate" is done, the code has to indicate explicitly when it's done using the register, so I think you'll a call to `cg->stopUsingRegister(bufReg)` at the end.

And again, if I understand correctly, you'll need to explicitly decrement the reference call for `node->getChild(0)` when you do a clobber evaluate on the node.

@0xdaryl, can you confirm that I have those requirements correct?

hzongaro replied:
> I do already decrement the reference count of all of node's children

Missed that.  Thanks!

dylanjtuttle commented:
Sounds good, I've squashed my commits!

hzongaro commented:
Jenkins test sanity.functional,sanity.openjdk xlinux,win,win32,xmac jdk8,jdk11,jdk17,jdk21

hzongaro commented:
Windows [JDK 11](https://openj9-jenkins.osuosl.org/job/Build_JDK11_x86-64_windows_Personal/1280/), [17](https://openj9-jenkins.osuosl.org/job/Build_JDK17_x86-64_windows_Personal/1156/) and [21](https://openj9-jenkins.osuosl.org/job/Build_JDK21_x86-64_windows_Personal/518/) build failures do not look like they could be related to this change.  I'm trying to reproduce the failures with a [fake pull request](https://github.com/eclipse-openj9/openj9/pull/22098), but the Windows builds there seem to be stalled.

hzongaro commented:
I was able to reproduce those Windows compile failures with my [fake pull request](https://github.com/eclipse-openj9/openj9/pull/22098), so they're definitely unrelated to these changes.

hzongaro commented:
Windows build failures were unrelated to this change.  Trying again.

Jenkins test sanity.functional,sanity.openjdk win,win32 jdk8,jdk11,jdk17,jdk21
